Plugins/Mute/DataManager.cs

```diff
@@ -1,38 +1,119 @@
+            client.ClientId);
+
+        if (clientMuteStateMeta is not null)
+        {
+            client.SetAdditionalProperty(Plugin.MuteKey, clientMuteStateMeta);
+            return clientMuteStateMeta;
+        }
+
+        // Get and return meta from old state.
+        var parseOldMuteState =
+            TryParse<MuteState>((await _metaService.GetPersistentMeta(Plugin.MuteKey, client.ClientId))?.Value,
+                out var oldClientMuteState);
+
+        // If still null, return client MuteStateMeta object. Not written.
+        // TODO: Ask Raid about this here... (Solves for nothing in DB and no old state (else a new joiner will trigger database read CONSTANTLY))
+        // TODO: If Raid agrees to keep this check, this Method will never return a nullable type.
+        if (!parseOldMuteState)
+        {
+            clientMuteStateMeta = new MuteStateMeta
+            {
+                ClientId = client.ClientId,
+                CleanedName = client.CleanedName,
+                Reason = string.Empty,
+                Expiration = DateTime.UtcNow,
+                AdminId = Utilities.IW4MAdminClient().ClientId,
+                MuteState = MuteState.Unmuted
+            };
+            client.SetAdditionalProperty(Plugin.MuteKey, clientMuteStateMeta);
+            return clientMuteStateMeta;
+        }
 
-        client.SetAdditionalProperty(Plugin.MuteKey, clientMuteState);
-        return clientMuteState;
+        // Populate meta so doesn't trigger infinite recursion in switch case.
+        // TODO: This sucks. I don't like it. It should be handled by MuteManager. Though, I don't know how.
+        clientMuteStateMeta = await WritePersistentData(client, Utilities.IW4MAdminClient(), DateTime.UtcNow,
+            MuteState.Unmuted, string.Empty);
+
+        switch (oldClientMuteState)
+        {
+            case MuteState.Muted:
+                // Run mute logic and update the return object.
+                await Plugin.MuteManager.Mute(client.CurrentServer, Utilities.IW4MAdminClient(), client,
+                    null, _translationLookup["PLUGINS_MUTE_MIGRATED"]);
+                clientMuteStateMeta = await ReadPersistentData(client);
+                break;
+            case MuteState.Unmuting:
+            case MuteState.Unmuted:
+            default:
+                break;
+        }
+
+        // Return null or the populated meta.
+        return clientMuteStateMeta;
     }
 
-    public async Task WritePersistentData(EFClient client, MuteState state)
+    public async Task<MuteStateMeta> WritePersistentData(EFClient client, EFClient adminId, DateTime? dateTime,
+        MuteState state, string reason)
     {
-        await _metaService.SetPersistentMeta(Plugin.MuteKey, state.ToString(), client.ClientId);
-        client.SetAdditionalProperty(Plugin.MuteKey, state);
+        var clientMuteStateMeta = client.GetAdditionalProperty<MuteStateMeta>(Plugin.MuteKey) ??
```

Comment:
The name of this method implies all data necessary to write should be provided as parameters. We should not need to look this up again before updating. Easiest thing to do is think of this doing a replacement of the value whether or not it exists already. Remember this is a state object we're writing so there should never really be a need for a partial update